util/types/datum_eval.go

```diff
@@ -395,3 +395,204 @@ func ComputeIntDiv(a, b Datum) (d Datum, err error) {
+func decimal2RoundUint(x *mysql.MyDecimal) (uint64, error) {
+	roundX := new(mysql.MyDecimal)
+	x.Round(roundX, 0)
+	var (
+		uintX uint64
+		err   error
+	)
+	if roundX.IsNegative() {
+		intX, err := roundX.ToInt()
+		if err != nil && err != mysql.ErrTruncated {
+			return 0, errors.Trace(err)
+		}
+		uintX = uint64(intX)
+	} else {
+		uintX, err = roundX.ToUint()
+		if err != nil && err != mysql.ErrTruncated {
+			return 0, errors.Trace(err)
+		}
+	}
+
+	return uintX, nil
+}
+
+// ComputeBitAnd computes the result of a & b.
+func ComputeBitAnd(a, b Datum) (d Datum, err error) {
+	aKind, bKind := a.Kind(), b.Kind()
+	if (aKind == KindInt64 || aKind == KindUint64) && (bKind == KindInt64 || bKind == KindUint64) {
+		d.SetUint64(a.GetUint64() & b.GetUint64())
+		return
+	}
+	// If either is not integer, we round the operands and then use uint64 to calculate.
+	x, err := a.ToDecimal()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	y, err := b.ToDecimal()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	uintX, err := decimal2RoundUint(x)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	uintY, err := decimal2RoundUint(y)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	d.SetUint64(uintX & uintY)
+	return d, nil
+}
+
+// ComputeBitOr computes the result of a | b.
+func ComputeBitOr(a, b Datum) (d Datum, err error) {
+	aKind, bKind := a.Kind(), b.Kind()
+	if (aKind == KindInt64 || aKind == KindUint64) && (bKind == KindInt64 || bKind == KindUint64) {
+		d.SetUint64(a.GetUint64() | b.GetUint64())
+		return
+	}
+	// If either is not integer, use decimal to calculate
+	x, err := a.ToDecimal()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	y, err := b.ToDecimal()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	uintX, err := decimal2RoundUint(x)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	uintY, err := decimal2RoundUint(y)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	d.SetUint64(uintX | uintY)
+	return d, nil
+}
+
+// ComputeBitNeg computes the result of ~a.
+func ComputeBitNeg(a Datum) (d Datum, err error) {
+	aKind := a.Kind()
+	if aKind == KindInt64 || aKind == KindUint64 {
+		d.SetUint64(^a.GetUint64())
+		return
+	}
+	// If either is not integer, we round the operands and then use uint64 to calculate.
+	x, err := a.ToDecimal()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	uintX, err := decimal2RoundUint(x)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	d.SetUint64(^uintX)
+	return d, nil
+}
+
+// ComputeBitXor computes the result of a ^ b.
+func ComputeBitXor(a, b Datum) (d Datum, err error) {
+	aKind, bKind := a.Kind(), b.Kind()
+	if (aKind == KindInt64 || aKind == KindUint64) && (bKind == KindInt64 || bKind == KindUint64) {
+		d.SetUint64(a.GetUint64() ^ b.GetUint64())
+		return
+	}
+	// If either is not integer, we round the operands and then use uint64 to calculate.
+	x, err := a.ToDecimal()
```

Comment:
Use a function to replace the following code block, then we could save a lot of works in ComputeXXX.